rdtools/test/soiling_test.py

```diff
@@ -224,3 +227,151 @@ def test_soiling_srr_with_nan_interval(normalized_daily, insolation, times):
+
+    pd.testing.assert_frame_equal(result, expected, atol=1e-8)
+
+# ###########################
+# monthly_soiling_rates tests
+# ###########################
+
+
+@pytest.fixture()
+def soiling_interval_summary():
+    starts = ['2019/01/01', '2019/01/16', '2019/02/08', '2019/03/06']
+    starts = pd.to_datetime(starts).tz_localize('America/Denver')
+    ends = ['2019/01/15', '2019/02/07', '2019/03/05', '2019/04/07']
+    ends = pd.to_datetime(ends).tz_localize('America/Denver')
+    slopes = [-0.005, -0.002, -0.001, -0.002]
+    slopes_low = [-0.0055, -0.0025, -0.0015, -0.003]
+    slopes_high = [-0.004, 0, 0, -0.001]
+    valids = [True, True, False, True]
+
+    soiling_interval_summary = pd.DataFrame()
+    soiling_interval_summary['start'] = starts
+    soiling_interval_summary['end'] = ends
+    soiling_interval_summary['slope'] = slopes
+    soiling_interval_summary['slope_low'] = slopes_low
+    soiling_interval_summary['slope_high'] = slopes_high
+    soiling_interval_summary['inferred_start_loss'] = np.nan
+    soiling_interval_summary['inferred_end_loss'] = np.nan
+    soiling_interval_summary['length'] = (ends - starts).days
+    soiling_interval_summary['valid'] = valids
+
+    return soiling_interval_summary
+
+
+def _build_monthly_summary(top_rows):
+    '''
+    Convienience function to build a full monthly soiling summary
+    dataframe from the expected_top_rows which summarize Jan-April
+    '''
+
+    all_rows = np.vstack((top_rows, [[1, np.nan, np.nan, np.nan, 0]]*8))
+
+    df = pd.DataFrame(data=all_rows,
+                      columns=['month', 'soiling_rate_median', 'soiling_rate_low', 'soiling_rate_high', 'interval_count'])
+    df['month'] = range(1, 13)
+
+    return df
+
+
+def test_monthly_soiling_rates(soiling_interval_summary):
+    np.random.seed(1977)
+    result = monthly_soiling_rates(soiling_interval_summary)
+
+    expected = np.array([[1.00000000e+00, -2.42103810e-03, -5.00912766e-03, -7.68551806e-04, 2.00000000e+00],
+                        [2.00000000e+00, -1.25092837e-03, -2.10091842e-03, -3.97354321e-04, 1.00000000e+00],
+                        [3.00000000e+00, -2.00313359e-03, -2.68359541e-03, -1.31927678e-03, 1.00000000e+00],
+                        [4.00000000e+00, -1.99729563e-03, -2.68067699e-03, -1.31667446e-03, 1.00000000e+00]])
+    expected = _build_monthly_summary(expected)
+
+    pd.testing.assert_frame_equal(result, expected, check_dtype=False)
+
+
+def test_monthly_soiling_rates_min_length(soiling_interval_summary):
+    np.random.seed(1977)
+    result = monthly_soiling_rates(soiling_interval_summary, min_length=20)
+
+    expected = np.array([[1.00000000e+00, -1.24851539e-03, -2.10394564e-03, -3.98358211e-04, 1.00000000e+00],
+                        [2.00000000e+00, -1.25092837e-03, -2.10091842e-03, -3.97330424e-04, 1.00000000e+00],
+                        [3.00000000e+00, -2.00309454e-03, -2.68359541e-03, -1.31927678e-03, 1.00000000e+00],
```

Comment:
I think your explanation is correct. I tested it by adding a `np.random.seed()` at line 852 in `soiling.py`. The same effect explains why there are slight differences between March and April as well, since only the same valid interval touches each.

Should we add more control for users to set the set within the monthly loop? I've tended to shy away from that because it is an extra parameter, and one can get repeatable results by setting the seeds the way we do in the tests.

Reply:
I could go either way on adding a seed parameter.  I'm a little skeptical that anyone would raise a fuss about a 0.002% difference, but who knows.  If it is added here in the name of reproducibility, probably something similar should be done elsewhere, no?

Reply:
Yeah let's leave it out for now. If people start clamoring for it we can always add it.